File: apps/api/app/models/booking.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.config import settings

from .base import Base


class Booking(Base):
    __tablename__ = "bookings"
    __table_args__ = {"schema": settings.db_schema}

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    booking_number: Mapped[str] = mapped_column(String(64), unique=True, nullable=False)

    deal_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey(f"{settings.db_schema}.deal_cards.id", ondelete="CASCADE"), nullable=False
    )
    practitioner_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey(f"{settings.db_schema}.practitioners.id", ondelete="CASCADE"), nullable=False
    )
    customer_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey(f"{settings.db_schema}.customers.id", ondelete="CASCADE"), nullable=False
    )

    customer_name: Mapped[str | None] = mapped_column(String(160), nullable=True)
    customer_email: Mapped[str] = mapped_column(String(255), nullable=False)
    customer_phone: Mapped[str | None] = mapped_column(String(64), nullable=True)
    avatar_url: Mapped[str | None] = mapped_column(String(500), nullable=True)

    quantity: Mapped[int] = mapped_column(default=1, nullable=False)
    subtotal: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    fee_amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False, default=Decimal("0.00"))
    total_amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    currency: Mapped[str] = mapped_column(String(8), nullable=False, default="USD")

    payment_status: Mapped[str] = mapped_column(String(32), nullable=False, default="pending")
    redemption_status: Mapped[str] = mapped_column(String(32), nullable=False, default="active")

    wallet_pass_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey(f"{settings.db_schema}.wallet_passes.id", ondelete="SET NULL"), nullable=True
    )
    qr_code: Mapped[str | None] = mapped_column(String(500), nullable=True)

    booked_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    redeemed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    refunded_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    deal = relationship("DealCard")
    practitioner = relationship("Practitioner")
    customer = relationship("Customer")
    wallet_pass = relationship("WalletPass", foreign_keys=[wallet_pass_id])

File: apps/web/src/domain/booking.ts
export type PaymentStatus = "pending" | "paid" | "refunded" | "failed";
export type RedemptionStatus = "active" | "redeemed" | "expired";

export type BookingApiPayload = {
  id: string;
  booking_number: string;
  deal_id: string;
  practitioner_id: string;
  customer_id: string;
  customer_name: string | null;
  customer_email: string;
  customer_phone: string | null;
  avatar_url: string | null;
  quantity: number;
  subtotal: string;
  fee_amount: string;
  total_amount: string;
  currency: string;
  payment_status: string;
  redemption_status: string;
  wallet_pass_id: string | null;
  qr_code: string | null;
  booked_at: string;
  redeemed_at: string | null;
  refunded_at: string | null;
  created_at: string;
  updated_at: string;
};

export type Booking = {
  id: string;
  booking_number: string;
  deal_id: string;
  practitioner_id: string;
  customer_id: string;
  customer_name: string | null;
  customer_email: string;
  customer_phone: string | null;
  avatar_url: string | null;
  quantity: number;
  subtotal: number;
  fee_amount: number;
  total_amount: number;
  currency: string;
  payment_status: PaymentStatus;
  redemption_status: RedemptionStatus;
  wallet_pass_id: string | null;
  qr_code: string | null;
  booked_at: string;
  redeemed_at: string | null;
  refunded_at: string | null;
  created_at: string;
  updated_at: string;
};

export function fromBookingApi(payload: BookingApiPayload): Booking {
  return {
    ...payload,
    subtotal: Number(payload.subtotal || 0),
    fee_amount: Number(payload.fee_amount || 0),
    total_amount: Number(payload.total_amount || 0),
    payment_status: normalizePaymentStatus(payload.payment_status),
    redemption_status: normalizeRedemptionStatus(payload.redemption_status)
  };
}

export function normalizePaymentStatus(value: string): PaymentStatus {
  if (value === "paid" || value === "refunded" || value === "failed") return value;
  return "pending";
}

export function normalizeRedemptionStatus(value: string): RedemptionStatus {
  if (value === "redeemed" || value === "expired") return value;
  return "active";
}

export function getPaymentStatusLabel(value: PaymentStatus): string {
  if (value === "paid") return "Paid";
  if (value === "refunded") return "Refunded";
  if (value === "failed") return "Failed";
  return "Pending";
}

export function getRedemptionStatusLabel(value: RedemptionStatus): string {
  if (value === "redeemed") return "Redeemed";
  if (value === "expired") return "Expired";
  return "Active";
}

export function getPaymentStatusColor(value: PaymentStatus): "green" | "red" | "amber" {
  if (value === "paid") return "green";
  if (value === "failed") return "red";
  return "amber";
}

export function getRedemptionStatusColor(value: RedemptionStatus): "green" | "red" | "amber" {
  if (value === "redeemed") return "green";
  if (value === "expired") return "red";
  return "amber";
}

export function formatBookingMoney(value: number, currency = "USD"): string {
  return new Intl.NumberFormat(undefined, {
    style: "currency",
    currency,
    minimumFractionDigits: 2,
    maximumFractionDigits: 2
  }).format(value);
}

export function formatBookingDate(value: string): string {
  return new Date(value).toLocaleString();
}
